### PR TITLE
feat(icon-button): Convert to TypeScript

### PR DESCRIPTION
**Pull request by abhiomkar**
Refs #4225

### CODE REVIEW

File: packages/mdc-icon-button/index.ts [hunk 1]
```diff
@@ -22,34 +22,25 @@
...
-  static attachTo(root) {
+class MDCIconButtonToggle extends MDCComponent<MDCIconButtonToggleFoundation> {
+  static attachTo(root: Element) {
     return new MDCIconButtonToggle(root);
   }
 
+  private ripple_: MDCRipple;
+  // TODO: type for event handler?
+  // tslint:disable-next-line:no-empty
+  private handleClick_ = (() => {});
```
Comment (moog16) [inline packages/mdc-icon-button/index.ts]: try this: 
```

import {SpecificEventListener} from '@material/dom/index';

...
private handleClick_: SpecificEventListener<'click'>;
```

File: packages/mdc-icon-button/index.ts [hunk 2]
```diff
@@ -22,34 +22,25 @@
...
+
+  // tslint:disable-next-line:no-any a component can pass in anything it needs to the constructor
```
Comment (moog16) [inline packages/mdc-icon-button/index.ts]: this comment is not disabling the next line for my ide

**Comment by moog16** [comment]
Don't forget to also update `js-bundle-factory.js` to 
```
iconButton: getAbsolutePath('/packages/mdc-icon-button/index.ts'),
```

and

File: packages/mdc-icon-button/adapter.ts [hunk 3]
```diff
@@ -38,38 +36,22 @@
  * the component in your framework of choice. See architecture documentation
  * for more details.
  * https://github.com/material-components/material-components-web/blob/master/docs/code/architecture.md
- *
```
[the remark this answers is not in the document]
Comment (kfranqueiro) [reply packages/mdc-icon-button/adapter.ts]: You mean remove the 2 lines above about closure compiler, right?

[the remark this answers is not in the document]
Comment (abhiomkar) [reply packages/mdc-icon-button/adapter.ts]: Thanks for spotting this. I'm also considering to remove _**...managing below things...**_ part.

File: packages/mdc-icon-button/index.ts (hunk 1 above)
Comment (abhiomkar) [reply packages/mdc-icon-button/index.ts]: that didn't work. used `EventListenerOrEventListenerObject` type instead.

File: packages/mdc-icon-button/adapter.ts [hunk 4]
```diff
@@ -21,55 +21,30 @@
  * THE SOFTWARE.
  */
 
-/* eslint no-unused-vars: [2, {"args": "none"}] */
-
 /**
- * Adapter for MDC Icon Button Toggle. Provides an interface for managing
- * - classes
- * - dom
```
Comment (moog16) [inline packages/mdc-icon-button/adapter.ts]: I think we can leave this list

File: packages/mdc-icon-button/foundation.ts [hunk 5]
```diff
@@ -22,37 +22,30 @@
...
+import {MDCIconButtonToggleAdapter} from './adapter';
 import {cssClasses, strings} from './constants';
 
-/**
- * @extends {MDCFoundation<!MDCIconButtonToggleAdapter>}
- */
-class MDCIconButtonToggleFoundation extends MDCFoundation {
-  static get cssClasses() {
+class MDCIconButtonToggleFoundation extends MDCFoundation<MDCIconButtonToggleAdapter> {
+  static get cssClasses(): {[key: string]: string} {
```
Comment (moog16) [inline packages/mdc-icon-button/foundation.ts]: Let the type be inferred 

https://github.com/material-components/material-components-web/pull/4319#discussion_r252748427

File: packages/mdc-icon-button/index.ts [hunk 6]
```diff
@@ -22,34 +22,27 @@
...
-  static attachTo(root) {
+class MDCIconButtonToggle extends MDCComponent<MDCIconButtonToggleFoundation> {
+  static attachTo(root: Element) {
     return new MDCIconButtonToggle(root);
   }
 
-  constructor(...args) {
-    super(...args);
+  private ripple_: MDCRipple;
+  private handleClick_: EventListenerOrEventListenerObject;
```
Comment (moog16) [inline packages/mdc-icon-button/index.ts]: Did the @material/Dom import not work?

**Comment by kfranqueiro** [comment]
Do we also need to update `index` to `index.ts` for icon-button in `material-components-web/index.js`? It looks like we've been doing that for the others as they've been converted.

File: packages/mdc-icon-button/index.ts [hunk 7]
```diff
@@ -22,34 +22,27 @@
...
-    /** @private {!MDCRipple} */
-    this.ripple_ = this.initRipple_();
-    /** @private {!Function} */
-    this.handleClick_;
-  }
+  constructor(
+    root: Element,
+    foundation?: MDCIconButtonToggleFoundation,
+    // tslint:disable-next-line:no-any a component can pass in anything it needs to the constructor
+    ...args: any[]
```
Comment (moog16) [inline packages/mdc-icon-button/index.ts]: I don't think we event need `...args`. Its only used in `mdc-base/component`'s `initialize` method. And there is not `initialize method.

File: packages/mdc-icon-button/index.ts (hunk 6 above)
Comment (moog16) [reply packages/mdc-icon-button/index.ts]: if you add the line below you can replace `EventListenerOrEventListenerObject ` --> `private handleClick_: SpecificEventListener<'click'>;`


```
root_!: HTMLElement;
```

**Comment by moog16** [comment]
@kfranquiero I think we may have missed a few component declarations in `material-components-web/index.js`. We can do it here to be consistent, but we will eventually remove all the `.ts` exts anyways during the clean up phase of https://github.com/material-components/material-components-web/issues/4225.

**Comment by kfranqueiro** [comment]
@moog16 afaict we have done it for every package that's been hit so far, with the exception of animation which doesn't appear in material-components-web/index.js at all. I was assuming it was necessary for the all-in-one build to work?

**Comment by moog16** [comment]
Don't think it is necessary since webpack resolves either `.js` or `.ts`.

File: packages/mdc-icon-button/index.ts (hunk 7 above)
Comment (acdvorak) [reply packages/mdc-icon-button/index.ts]: We probably shouldn't override the constructor. Instead, initialize the properties in `initialSyncWithDOM()` and mark them with `!`:

```js
  private ripple_!: MDCRipple;
  private handleClick_!: EventListenerOrEventListenerObject;
```

**Pull request merged**